### ISSUE TITLE
Installation can be quite an ordeal (sudo | sudo --ignore-scripts and update | sudo --user=$(whoami))

### ISSUE THREAD

**Issue by LukasRypl** [tldr-pages/tldr-node-client#123]
There are at least 4 ways to install tldr (as normal user, via sudo, with --ignore-scripts and manual update, via sudo --user=$(whoami)). Not sure which should be used because none seems to be working out of the box in my case.

It looks to me as if installation script:
1. expects that the directory `~/.tldr` exists already
2. does not cleanup `/tmp/tldr` and `~/.tldr` in case of failure and upon uninstalling tldr

These things combined can turn a simple installation into quite a tedious task...

## Expected behavior
Install and use tldr node client :)

## Actual behavior
 
1. Installed npm via package manager.
1. Tried to install tldr as a user (because it is written here https://tldr-pages.github.io/#installation) -> `npm install tldr -g` failed
1. Tried again via sudo because it is what https://github.com/tldr-pages/tldr-node-client#installing says  `sudo npm install tldr -g` -> failed
1. Found a workaround mentioned somewhere on github, tried again `npm install -g --ignore-scripts tldr` -> failed
1. try again via sudo  `sudo npm install -g --ignore-scripts tldr` -> installation is OK 👍 
1. `tldr --update` -> failed
1. `sudo tldr --update` -> failed (not a real problem)
1. `sudo /usr/local/bin/tldr --update` looks OK
1. `tldr ls` -> please update 
Failed to update and started writing this description ...
1. `mkdir -p ~/.tldr` to resolve one obvious problem
1. `tldr --update` -> failed 
1. `ls -l /tmp/tldr` so there was something left  -> `sudo rm -r /tmp/tldr` 
1. `tldr --update` 👍 
1. `tldr ls` 👍 👍 
Finally it works :)

Discovered that another suggested installation command is described at https://github.com/tldr-pages/tldr#tldr -- well it looks quite nice so I tried again:
1. remove package as root `# npm rm -g tldr`
1. cleanup `# rm -r /root/.tldr`
1. `# rm -r /tmp/tldr`
1. as a user `~> rm -r ~/.tldr`
1. `sudo npm install -g tldr --user=$(whoami)` -> failed again 

## Log, debug output

```
~> npm install tldr -g
npm ERR! Linux 4.4.36-8-default
npm ERR! argv "/usr/bin/node" "/usr/bin/npm" "install" "tldr" "-g"
npm ERR! node v4.6.1
npm ERR! npm  v2.15.9
npm ERR! path /usr/local/lib
npm ERR! code EACCES
npm ERR! errno -13
npm ERR! syscall mkdir

npm ERR! Error: EACCES: permission denied, mkdir '/usr/local/lib'
npm ERR!     at Error (native)
npm ERR!  { [Error: EACCES: permission denied, mkdir '/usr/local/lib']
npm ERR!   errno: -13,
npm ERR!   code: 'EACCES',
npm ERR!   syscall: 'mkdir',
npm ERR!   path: '/usr/local/lib' }
npm ERR! 
npm ERR! Please try running this command again as root/Administrator.

npm ERR! Please include the following file with any support request:
npm ERR!     /home/rypl/npm-debug.log

~> sudo npm install tldr -g
root's password:
npm WARN deprecated node-uuid@1.4.7: use uuid module instead
/usr/local/bin/tldr -> /usr/local/lib/node_modules/tldr/bin/tldr

> tldr@1.7.0 postinstall /usr/local/lib/node_modules/tldr
> node ./bin/tldr --update

Updating...
/usr/local/lib/node_modules/tldr/node_modules/fs-extra/lib/mkdirs/mkdirs-sync.js:47
          throw err0
          ^

Error: EACCES: permission denied, mkdir '/root/.tldr/cache'
    at Error (native)
    at Object.fs.mkdirSync (fs.js:794:18)
    at Object.mkdirsSync (/usr/local/lib/node_modules/tldr/node_modules/fs-extra/lib/mkdirs/mkdirs-sync.js:29:9)
    at /usr/local/lib/node_modules/tldr/lib/cache.js:36:8
    at Extract.<anonymous> (/usr/local/lib/node_modules/tldr/lib/remote.js:27:5)
    at emitNone (events.js:72:20)
    at Extract.emit (events.js:166:7)
    at DirWriter.<anonymous> (/usr/local/lib/node_modules/tldr/node_modules/unzip2/lib/extract.js:42:10)
    at emitNone (events.js:72:20)
    at DirWriter.emit (events.js:166:7)
npm ERR! Linux 4.4.36-8-default
npm ERR! argv "/usr/bin/node" "/usr/bin/npm" "install" "tldr" "-g"
npm ERR! node v4.6.1
npm ERR! npm  v2.15.9
npm ERR! code ELIFECYCLE

npm ERR! tldr@1.7.0 postinstall: `node ./bin/tldr --update`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the tldr@1.7.0 postinstall script 'node ./bin/tldr --update'.
npm ERR! This is most likely a problem with the tldr package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node ./bin/tldr --update
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs tldr
npm ERR! Or if that isn't available, you can get their info via:
npm ERR! 
npm ERR!     npm owner ls tldr
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /home/rypl/npm-debug.log

 ~> npm install -g --ignore-scripts tldr
npm ERR! Linux 4.4.36-8-default
npm ERR! argv "/usr/bin/node" "/usr/bin/npm" "install" "-g" "--ignore-scripts" "tldr"
npm ERR! node v4.6.1
npm ERR! npm  v2.15.9
npm ERR! path /usr/local/lib
npm ERR! code EACCES
npm ERR! errno -13
npm ERR! syscall mkdir

npm ERR! Error: EACCES: permission denied, mkdir '/usr/local/lib'
npm ERR!     at Error (native)
npm ERR!  { [Error: EACCES: permission denied, mkdir '/usr/local/lib']
npm ERR!   errno: -13,
npm ERR!   code: 'EACCES',
npm ERR!   syscall: 'mkdir',
npm ERR!   path: '/usr/local/lib' }
npm ERR! 
npm ERR! Please try running this command again as root/Administrator.

npm ERR! Please include the following file with any support request:
npm ERR!     /home/rypl/npm-debug.log

~> sudo npm install -g --ignore-scripts tldr
npm WARN deprecated node-uuid@1.4.7: use uuid module instead
/usr/local/bin/tldr -> /usr/local/lib/node_modules/tldr/bin/tldr
tldr@1.7.0 /usr/local/lib/node_modules/tldr
├── lodash.isempty@4.4.0
├── lodash.sample@4.2.1
├── lodash.last@3.0.0
├── lodash.unescape@4.0.1
├── lodash.memoize@4.1.2
├── lodash.compact@3.0.1
├── os-homedir@1.0.2
├── lodash.identity@3.0.0
├── ms@0.7.2
├── lodash.defaults@4.2.0
├── lodash.get@4.4.2
├── lodash.partial@4.2.1
├── lodash.map@4.6.0
├── marked@0.3.6
├── commander@2.9.0 (graceful-readlink@1.0.1)
├── chalk@1.1.3 (escape-string-regexp@1.0.5, supports-color@2.0.0, ansi-styles@2.2.1, has-ansi@2.0.0, strip-ansi@3.0.1)
├── request@2.75.0 (aws-sign2@0.6.0, tunnel-agent@0.4.3, forever-agent@0.6.1, oauth-sign@0.8.2, caseless@0.11.0, is-typedarray@1.0.0, stringstream@0.0.5, aws4@1.5.0, isstream@0.1.2, json-stringify-safe@5.0.1, extend@3.0.0, qs@6.2.1, node-uuid@1.4.7, combined-stream@1.0.5, mime-types@2.1.13, form-data@2.0.0, tough-cookie@2.3.2, hawk@3.1.3, bl@1.1.2, http-signature@1.1.1, har-validator@2.0.6)
├── fs-extra@0.30.0 (path-is-absolute@1.0.1, klaw@1.3.1, jsonfile@2.4.0, graceful-fs@4.1.11, rimraf@2.5.4)
└── unzip2@0.2.5 (setimmediate@1.0.5, pullstream@0.4.1, readable-stream@1.0.34, match-stream@0.0.2, binary@0.3.0, fstream@0.1.31)

 ~> tldr ls
Page not found
Try updating with "tldr --update", or submit a pull request to:
https://github.com/tldr-pages/tldr

 ~> tldr --update
Updating...
/usr/local/lib/node_modules/tldr/node_modules/fs-extra/node_modules/rimraf/rimraf.js:312
        throw er
        ^

Error: EACCES: permission denied, unlink '/tmp/tldr/LICENSE.md'
    at Error (native)
    at Object.fs.unlinkSync (fs.js:932:18)
    at rimrafSync (/usr/local/lib/node_modules/tldr/node_modules/fs-extra/node_modules/rimraf/rimraf.js:305:17)
    at /usr/local/lib/node_modules/tldr/node_modules/fs-extra/node_modules/rimraf/rimraf.js:340:5
    at Array.forEach (native)
    at rmkidsSync (/usr/local/lib/node_modules/tldr/node_modules/fs-extra/node_modules/rimraf/rimraf.js:339:26)
    at rmdirSync (/usr/local/lib/node_modules/tldr/node_modules/fs-extra/node_modules/rimraf/rimraf.js:332:7)
    at Function.rimrafSync [as sync] (/usr/local/lib/node_modules/tldr/node_modules/fs-extra/node_modules/rimraf/rimraf.js:303:9)
    at Object.removeSync (/usr/local/lib/node_modules/tldr/node_modules/fs-extra/lib/remove/index.js:4:17)
    at Object.exports.download (/usr/local/lib/node_modules/tldr/lib/remote.js:19:6)

 ~> sudo tldr --update
sudo: tldr: command not found

 ~> sudo /usr/local/bin/tldr --update
Updating...
Done

 ~> tldr ls
Page not found
Try updating with "tldr --update", or submit a pull request to:
https://github.com/tldr-pages/tldr

~> mkdir -p ~/.tldr

~> tldr --update
Updating...
/usr/local/lib/node_modules/tldr/node_modules/fs-extra/node_modules/rimraf/rimraf.js:312
        throw er
        ^

Error: EACCES: permission denied, unlink '/tmp/tldr/LICENSE.md'
    at Error (native)
    at Object.fs.unlinkSync (fs.js:932:18)
    at rimrafSync (/usr/local/lib/node_modules/tldr/node_modules/fs-extra/node_modules/rimraf/rimraf.js:305:17)
    at /usr/local/lib/node_modules/tldr/node_modules/fs-extra/node_modules/rimraf/rimraf.js:340:5
    at Array.forEach (native)
    at rmkidsSync (/usr/local/lib/node_modules/tldr/node_modules/fs-extra/node_modules/rimraf/rimraf.js:339:26)
    at rmdirSync (/usr/local/lib/node_modules/tldr/node_modules/fs-extra/node_modules/rimraf/rimraf.js:332:7)
    at Function.rimrafSync [as sync] (/usr/local/lib/node_modules/tldr/node_modules/fs-extra/node_modules/rimraf/rimraf.js:303:9)
    at Object.removeSync (/usr/local/lib/node_modules/tldr/node_modules/fs-extra/lib/remove/index.js:4:17)
    at Object.exports.download (/usr/local/lib/node_modules/tldr/lib/remote.js:19:6)

 ~> ls -lh /tmp/tldr/
total 4.0K
-rw-r--r-- 1 root root 1.2K Dec 29 21:56 LICENSE.md
drwxr-xr-x 1 root root   58 Dec 29 21:56 pages/

 ~> tldr --update
Updating...
Done

 ~> tldr ls
  ls

  List directory contents.

  - List files one per line:
    ls -1
    ....
```

## Environment
- Operating system - Linux openSUSE Leap 42.2
- NodeJS version `zypper info nodejs4`
...
Repository     : Main Update Repository       
Name           : nodejs4                      
Version        : 4.6.1-3.1                    
Arch           : x86_64                       
Vendor         : openSUSE                     
Installed Size : 12.7 MiB                     
Installed      : Yes                          
Status         : up-to-date                   

---

Hope it helps :)

**Comment by LukasRypl** [comment]
Is it solved in #119 ?

**Comment by agnivade** [comment]
Yep :) Thanks a lot for your detailed report. Sorry that you had to go through all that trouble to install it. I know this has been a pain point for long. I just pushed the fix yesterday and will do a minor point release today.

Just one more thing - You mention installing with `sudo` and as normal user as two different ways. I don't think those can be termed as different per se. Any command can be prepended with a sudo. But that doesn't create a different command. In this case, depending on the permissions of the user's `/usr/local/lib`, one adds or does not add `sudo`.

And please let me know where you found the `--ignore-scripts` method. Its an old one which was superceded by the `--user` option.

**Comment by agnivade** [comment]
Just published the 1.7.1 release. Closing the issue.

**Comment by LukasRypl** [comment]
Thanks, 1.7.1 works like a charm.

> Just one more thing - You mention installing with sudo and as normal user as two different ways.

You are right, the commands are the same, only the environment differ. In this particular case the path to the home directory is important.


The `--ignore-scripts` was mentioned in discussion https://github.com/tldr-pages/tldr-node-client/issues/91#issuecomment-191287855 (March 2016).
